Lib/concurrent/futures/process.py

```diff
@@ -435,6 +435,10 @@ def shutdown_worker():
                 # is not gc-ed yet.
                 if executor is not None:
                     executor._shutdown_thread = True
+                    if executor._cancel_pending_work:
+                        for work_item in pending_work_items.values():
+                            work_item.future.cancel()
```

Comment:
How does this interact with the next line? Would it be safe to do `pending_work_items.clear()`?

Reply:
I tried to do that at first, but it causes issues with futures that just started running; they end up being not able to finish. I believe this is due to the implementation of `shutdown_worker()`. Specifically, a work item is considered to be "running" once it's moved from the pending work items dict into the call queue in `_add_item_to_call_queue()`. But, there's a period when it sits in the call queue briefly before being executed. This gets interrupted in `shutdown_worker()` when it pushes `None` to the call queue.

The end result of this is that the running futures are left in a state of "PENDING" after `executor.shutdown()` returns, even when `wait=True` (which is not the intended behavior)`.

Reply:
Also, just to double check, I attempted to add the following change:
```py
                     if executor._cancel_pending_work:
                        for work_item in pending_work_items.values():
                            work_item.future.cancel()
+                       pending_work_items.clear()
```
This resulted in the following test results:

```
[aeros:~/repos/aeros-cpython]$ ./python -m test test_concurrent_futures --match test_cancel_futures -v
== CPython 3.9.0a2+ (heads/bpo37224-fix-test__xxsubinterpreters.DestroyTests:6cbd9919ed, Jan 24 2020,) [GCC 9.2.0]
== Linux-5.4.8-arch1-1-x86_64-with-glibc2.30 little-endian
== cwd: /home/aeros/repos/aeros-cpython/build/test_python_156556
== CPU count: 4
== encodings: locale=UTF-8, FS=utf-8
0:00:00 load avg: 0.12 Run tests sequentially
0:00:00 load avg: 0.12 [1/1] test_concurrent_futures
test_cancel_futures (test.test_concurrent_futures.ProcessPoolForkProcessPoolShutdownTest) ... 0.31s FAIL
test_cancel_futures (test.test_concurrent_futures.ProcessPoolForkserverProcessPoolShutdownTest) ... 0.22s FAIL
test_cancel_futures (test.test_concurrent_futures.ProcessPoolSpawnProcessPoolShutdownTest) ... 0.32s FAIL
test_cancel_futures (test.test_concurrent_futures.ThreadPoolShutdownTest) ... 0.20s ok

======================================================================
FAIL: test_cancel_futures (test.test_concurrent_futures.ProcessPoolForkProcessPoolShutdownTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/aeros/repos/aeros-cpython/Lib/test/test_concurrent_futures.py", line 363, in test_cancel_futures
    self.assertTrue(fut.done())
AssertionError: False is not true

======================================================================
FAIL: test_cancel_futures (test.test_concurrent_futures.ProcessPoolForkserverProcessPoolShutdownTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/aeros/repos/aeros-cpython/Lib/test/test_concurrent_futures.py", line 363, in test_cancel_futures
    self.assertTrue(fut.done())
AssertionError: False is not true

======================================================================
FAIL: test_cancel_futures (test.test_concurrent_futures.ProcessPoolSpawnProcessPoolShutdownTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/aeros/repos/aeros-cpython/Lib/test/test_concurrent_futures.py", line 363, in test_cancel_futures
    self.assertTrue(fut.done())
AssertionError: False is not true

----------------------------------------------------------------------

Ran 4 tests in 1.089s

FAILED (failures=3)
test test_concurrent_futures failed
test_concurrent_futures failed

== Tests result: FAILURE ==

1 test failed:
    test_concurrent_futures

Total duration: 1.2 sec
Tests result: FAILURE
```

Reply:
Do you understand why? I don't want to introduce any race conditions or other concurrency issues?

Reply:
> Do you understand why?

I'll have to do some further investigation to be 100% certain, but I believe that clearing the pending items dict leads to issues because of the implementation of `shutdown_worker()`.

Specifically, in `shutdown_worker()` it sends a wake-up to the call queue (via `call_queue.put_nowait(None)`), which interrupts any running workers. This results in the futures being left in a running state instead of being allowed to complete (even when `shutdown()` is called w/ `wait=True`).

Once I get the chance (probably later tonight or tomorrow), I'd like to test moving this section to *after* the block for `if not pending_work_items`, and see how that behaves. This might allow us to still clear the pending items dict after cancelling the pending futures, without interrupting the futures that are currently running. The only minor downside is that it will require a second `if executor is not None` conditional before cancelling the futures (to check `if executor._cancel_pending_work`), but that's a very small (if not entirely negligible) performance cost.

I'll report back with any results.

Reply:
@brianquinlan While investigating the above, I discovered the following:

1) The futures in `pending_work_items` are actually a mix of `PENDING` and `RUNNING` futures, not just `PENDING` futures as I had initially assumed. I hadn't realized that futures will left in `pending_work_items` after being set to `RUNNING` in `_add_call_item_to_queue()` until they were finished and had a result assigned. With that in mind, I think the best way to handle this part would be to attempt to cancel all of the futures with `if not work_item.future.cancel()`, adding any `RUNNING` futures to a new dict that will replace the current `pending_work_items` to remove the `PENDING` futures (since we can't mutate a dict in-place while iterating through it). This means that we won't call `shutdown_worker()` until all of the running futures have finished, which is the intended behavior. 

We should also drain the `work_ids` queue, since we're no longer adding items to the call queue at this point. Otherwise, we'd have to modify `_add_call_item_to_queue()` to use `pending_work_items.get(work_id, None)` instead of the current `pending_work_items[work_id]`, which would be a slight performance loss even when not using `cancel_futures`. I'd prefer to avoid that if possible.

2) Some of the logic in my `test_cancel_futures` was a bit flawed. Specifically, using `time.sleep(.1)` in between the `submit()` calls and the `shutdown()` call leaves the test far too vulnerable to race conditions. Also, I found that there's no direct connection between the max workers and the number of workers that will immediately start running, so we definitely shouldn't try to guarantee that. It was likely just a coincidence that it even worked before, probably as a result of the `time.sleep(.1)` before shutting the executor down. I'll fix the tests accordingly.

One part of the tests I'd like to make a bit more constraining is the number of cancelled futures. From observing it's behavior (when running up to 200 parallel test jobs), I think at least 25 of the 50 submitted futures should be cancelled here to ensure `cancel_futures=True` is working as intended. The average seemed to be around ~40-46 (depending on TPE vs PPE, method of process spawning, number of workers, and CPU load), but I'd like to give enough room for flexibility on different platforms.